Restore the bogus carrier / calculator to unit test active_shipping/b…

fredericboivin opened the pull request:
active_shipping_calculator_spec.rb is in a weird state since it's testing with a USPS calculators, who decorates the base methods like find_rates from active_shipping/base.rb

So this PR is to : 

1) Restore the use of bogus carrier / calculator in the active_shipping_calculator_spec to unit tests the code present in active_shipping/base.rb

2) Extract the USPS specific code into a new carrier spec as a first step

fredericboivin committed:
Extract USPS specific code to a new files

- Enable most xit tests and make them green (the weight related ones are still xit)
- Do a basic cleanup of the syntax to use expect instead of should (to be continued)
- Extract the compute test to a new carrier/USPS file
- Restore usage of Bogus carrier / Bogus calculator to only unit test active_shipping/base.rb

jhawthorn commented:
LGTM :+1:

forkata commented on spec/models/active_shipping_calculator_spec.rb:
```diff
@@ -108,11 +106,11 @@ module ActiveShipping
-        stub_request(:get, /http:\/\/production.shippingapis.com\/ShippingAPI.dll.*/).
-          to_return(:body => fixture(:normal_rates_request))
-        calculator.compute(package).should == 14.1
+        carrier.should_receive(:find_rates).and_return(ActiveShipping::RateResponse.new(true, "Foo"))
```
It would be nice if we can start moving away from the older `should_receive` syntax to the newer `expect(foo).to receive`.

forkata replied:
On second thought, the `BogusCarrier` should already return a fake `RateResponse` so you might not even need to stub this here. See https://github.com/solidusio-contrib/solidus_active_shipping/blob/master/spec/lib/spree/active_shipping/bogus_carrier.rb#L11

forkata commented on spec/models/active_shipping_calculator_spec.rb:
```diff
@@ -126,31 +124,35 @@ module ActiveShipping
+        end
+
         before do
           carrier.should_receive(:find_rates).and_return(response)
```
Would prefer if this used the new rspec syntaxt `allow(carrier).to receive(:find_rates)`.

forkata commented on spec/models/active_shipping_calculator_spec.rb:
```diff
@@ -126,31 +124,35 @@ module ActiveShipping
-        xit "should return rate based on calculator's service_name" do
-          calculator.class.should_receive(:description).and_return("Super Fast")
+        it "should return rate based on calculator's service_name" do
+          calculator.class.should_receive(:description).and_return("Bogus Calculator")
```
Same as above, would prefer `allow` syntax.

forkata commented on spec/models/active_shipping_calculator_spec.rb:
```diff
@@ -126,31 +124,35 @@ module ActiveShipping
-        xit "should include handling_fee when configured" do
-          calculator.class.should_receive(:description).and_return("Super Fast")
+        it "should include handling_fee when configured" do
+          calculator.class.should_receive(:description).and_return("Bogus Calculator")
```
Same as above, would prefer `allow` syntax.

forkata commented on spec/models/carriers/usps_calculator_spec.rb:
```diff
@@ -0,0 +1,97 @@
+    let(:package) { order.shipments.first.to_package }
+
+    before(:each) do
+      Spree::StockLocation.destroy_all
```
If we create the stock location before the order and line items, can we get around doing this. I would try making that a `let!(:stock_location)` and see if it fixes this.

forkata commented on spec/models/carriers/usps_calculator_spec.rb:
```diff
@@ -0,0 +1,97 @@
+      Spree::StockLocation.destroy_all
+      stock_location
+      order.create_proposed_shipments
+      order.shipments.count.should == 1
```
I would prefer if we used the newer `expect(order.shipments.count).to eq 1` here.

forkata commented on spec/models/active_shipping_calculator_spec.rb:
```diff
@@ -108,11 +106,11 @@ module ActiveShipping
 
     describe "compute" do
```
Would be nice if we can define the `subject` for this block to be `subject { calculator.compute(package) }`. That way we can just call `subject` or use the `it { is_expected.to ... }` syntax.

forkata commented on spec/models/carriers/usps_calculator_spec.rb:
```diff
@@ -0,0 +1,97 @@
+      it{ expect(package.stock_location).to eq(stock_location) }
+    end
+
+    describe "compute" do
```
Same as above, we can define the subject for this block to be `subject { calculator.compute(package) }`.

forkata commented on spec/models/carriers/usps_calculator_spec.rb:
```diff
@@ -0,0 +1,97 @@
+      it "should use the carrier supplied in the initializer" do
+        stub_request(:get, /http:\/\/production.shippingapis.com\/ShippingAPI.dll.*/).
+            to_return(:body => fixture(:normal_rates_request))
+        calculator.compute(package).should == 14.1
```
Would prefer `expect(...).to eq` syntax.

forkata commented:
@dangerdogz This is a great start for improving the tests. Aside from the few style comments I am :+1: on the direction for this.

fredericboivin commented:
Good feeback. I share the style concerns and I'll get started on cleaning up with the new syntax

fredericboivin committed:
Modernize Rspec syntax

- Use rubocop to cleanup the hash syntax, spacing etc.
- Continue the work of modernizing the Rspec syntax (expect, allow)

fredericboivin commented:
I did a first pass on the active_shipping_spec, still need to sort the stock_location part but otherwise it looks better. I very much enjoy the process since learning to write good tests is definitely where I have the most work to do. (also cleaned it up a little with rubocop)

forkata commented:
@dangerdogz That looks great. Thanks so much for all your work on this!

fredericboivin commented on spec/lib/spree/active_shipping/bogus_carrier.rb:
```diff
@@ -8,7 +8,7 @@ def name
 
       def find_rates(origin, destination, packages, options = {})
-        RateResponse.new(true, "Bogus rate response.")
+        ::ActiveShipping::RateResponse.new(true, "Bogus response")
```
the ::ActiveShipping is not needed after all, since it's derived from ::ActiveShipping::Carrier already

EDIT ( I lied it's needed, it's not crashing because it's never called)

Senjai commented on spec/models/carriers/usps_calculator_spec.rb:
```diff
@@ -0,0 +1,95 @@
+require 'spec_helper'
+
+module ActiveShipping
```
I'm unsure why this spec has to live in this module.

Senjai commented on spec/models/carriers/usps_calculator_spec.rb:
```diff
@@ -0,0 +1,95 @@
+
+module ActiveShipping
+  describe Spree::Calculator::Shipping::Usps do
+    WebMock.disable_net_connect!
```
Is this usually enabled? If so do we need to turn it back on after tests?

Either way we should probably do this in a before

Senjai commented on spec/models/carriers/usps_calculator_spec.rb:
```diff
@@ -0,0 +1,95 @@
+    let(:package) { order.shipments.first.to_package }
+
+    before(:each) do
+      Spree::StockLocation.destroy_all
```
If we rebase on top of the database cleaner dealy does this go away?

fredericboivin replied:
Yes if you use let! with the stock_location (I'm waiting on the other pull request being merged to finish cleaning this up)

Senjai commented on spec/models/carriers/usps_calculator_spec.rb:
```diff
@@ -0,0 +1,95 @@
+      Spree::ActiveShipping::Config.set(units: 'imperial')
+      Spree::ActiveShipping::Config.set(unit_multiplier: 1)
+      allow(calculator).to receive(:carrier) { carrier }
+      Rails.cache.clear
```
Why? (Probably because of the cached shipping results for packages dealy, but we need a comment none the less, most people won't know this).

Even so, if we can Rails.cache.delete(cache_key(package)), that might be better.

fredericboivin replied:
Note : This part is a bit "leaky"

If you disable the cache, sometimes the compute spec will fail because  

``` ruby
it "should return rate based on calculator's service_name" do
    allow(calculator.class).to receive(:description) { 'Bogus Calculator' }
    expect(subject).to eq 9.99
end
```

will receive the cached results from :

``` ruby
it 'should include handling_fee when configured' do
  Spree::ActiveShipping::Config.set(handling_fee: 100)
   allow(calculator.class).to receive(:description) { 'Bogus Calculator' }
   expect(subject).to eq 10.99
end
```

It's expecting 9.99 and it gets 10.99 from the cache
So it's potentially a bug though

For now, I will explicitly set no fees in the before(:each), and set them to 1.00 in the test where it's needed.

Senjai commented on spec/models/carriers/usps_calculator_spec.rb:
```diff
@@ -0,0 +1,95 @@
+      Spree::StockLocation.destroy_all
+      stock_location
+      order.create_proposed_shipments
+      expect(order.shipments.count).to eq 1
```
We're not testing `Spree::Order#create_proposed_shipments`

fredericboivin replied:
You mean that it's not necessary to "expect(order.shipments.count).to eq 1" since we're not testing create_proposed_shipments?

Senjai commented on spec/models/carriers/usps_calculator_spec.rb:
```diff
@@ -0,0 +1,95 @@
+
+      context 'with valid response' do
+        let(:rates) do
+          [double('rate', service_code: '3', price: 999)]
```
We should use an instance_double here to verify the methods, or preferably no stubbing at all given we're already stubbing the request.

Senjai commented on spec/models/carriers/usps_calculator_spec.rb:
```diff
@@ -0,0 +1,95 @@
+        line_item.quantity = 2
+        line_item.variant.save
+        line_item.variant.weight = 1
+        line_item.variant.save
```
j/c, why do we have to save after every assignment?

Senjai commented on spec/models/carriers/usps_calculator_spec.rb:
```diff
@@ -0,0 +1,95 @@
+
+    let(:carrier) { ActiveShipping::USPS.new(login: 'FAKEFAKEFAKE') }
+    let(:calculator) { Spree::Calculator::Shipping::Usps::ExpressMail.new }
+    let(:response) { double('response', rates: rates, params: {}) }
```
instance_double

Senjai commented on spec/models/carriers/usps_calculator_spec.rb:
```diff
@@ -0,0 +1,95 @@
+        expect(subject).to eq 14.1
+      end
+
+      context 'with valid response' do
```
contexts should come in pairs. If we see a "with valid response", I would expect a "with an invalid response"

fredericboivin committed:
Remove the StockLocation.destroy before each test

By using database cleaner and let! we only need to create it once

fredericboivin committed:
Refactor the Order creation through FactoryGirl

Avoid manually editing and saving variants between each tests, instead we can
use the factories to create them properly through the Order only once.

fredericboivin committed:
Remove unneeded namespacing

- Remove the module ActiveShipping wrapping the tests
- Remove the Webmock.disable from the tests (not needed because
we're using a stubbed request)

fredericboivin committed:
Explicitly set handling fees

Explicitly set handling fees so we can avoid spurious failures that happens when
cached rates with handling fees are returned from cache for tests
without handling fees.

Delete only the cache keys we're adding instead of the whole cache before each tests.

fredericboivin committed:
Remove the stubbed response

- Remove the stubbing around the calculator for the USPS suite since we're using
a real carrier and a real calculator.
- Replace the response stubbing with a fixture from a live call

fredericboivin commented:
I've cleaned up the stub usage in the usps spec since it was copy pasted from active_shipping_calculator_spec.

Since we're using a real calculator with a real carrier, the only thing we need to stub is the HTTP call since we're using a fixture.

fredericboivin committed:
Remove shipments count test

Remove the shipment count expectation in the before block. We are not
testing create_proposed_shipments and it is out of scope for this file.

Commit:
Add test coverage for failed ActiveShipping calls

Add a context for ActiveShipping returning a ResponseError. Expected behavior
is a raise Spree::ShippingError

Senjai commented on app/models/spree/calculator/shipping/usps/base.rb:
```diff
@@ -51,7 +51,6 @@ def retrieve_rates(origin, destination, shipment_packages)
             rate_hash = Hash[*rates.flatten]
             return rate_hash
           rescue ::ActiveShipping::Error => e
-
```
Mistype?

Senjai commented on spec/models/active_shipping_calculator_spec.rb:
```diff
@@ -1,167 +1,170 @@
+
+  let(:carrier) { Spree::ActiveShipping::BogusCarrier.new }
+  let(:calculator) { Spree::Calculator::Shipping::ActiveShipping::BogusCalculator.new }
+  let(:response) { double('response', rates: rates, params: {}) }
```
Can we instance double this?

Senjai commented on spec/models/active_shipping_calculator_spec.rb:
```diff
@@ -1,167 +1,170 @@
+  describe 'available' do
+    context 'when rates are available' do
+      let(:rates) do
+        [double('rate', service_name: 'Bogus Calculator', price: 1)]
```
Qualified doubles please.

Senjai commented:
:+1:
